x-pack/plugins/ml/server/saved_objects/service.ts

```diff
@@ -366,6 +406,335 @@ export function jobSavedObjectServiceFactory(
+          await savedObjectsClient.delete(ML_TRAINED_MODEL_SAVED_OBJECT_TYPE, modelId, {
+            force: true,
+          });
+        }
+      }
+    } catch (error) {
+      // the saved object may exist if a previous job with the same ID has been deleted.
+      // if not, this error will be throw which we ignore.
+    }
+    let initialNamespaces;
+    // if a job exists for this model, ensure the initial namespaces for the model
+    // are the same as the job
+    if (job !== null) {
+      const [existingJobObject] = await getAllJobObjectsForAllSpaces(
+        'data-frame-analytics',
+        job.job_id
+      );
+
+      initialNamespaces = existingJobObject?.namespaces ?? undefined;
+    }
+
+    await savedObjectsClient.create<TrainedModelObject>(
+      ML_TRAINED_MODEL_SAVED_OBJECT_TYPE,
+      modelObject,
+      {
+        id: modelId,
+        ...(initialNamespaces ? { initialNamespaces } : {}),
+      }
+    );
+  }
+
+  async function _bulkCreateTrainedModel(models: TrainedModelObject[], namespaceFallback?: string) {
+    await isMlReady();
+
+    const namespacesPerJob = (await getAllJobObjectsForAllSpaces()).reduce((acc, cur) => {
+      acc[cur.attributes.job_id] = cur.namespaces;
+      return acc;
+    }, {} as Record<string, string[] | undefined>);
+
+    return await savedObjectsClient.bulkCreate<TrainedModelObject>(
+      models.map((m) => {
+        let initialNamespaces = m.job && namespacesPerJob[m.job.job_id];
+        if (!initialNamespaces?.length && namespaceFallback) {
+          // use the namespace fallback if it is defined and no namespaces can
+          // be found for a related job.
+          // otherwise initialNamespaces will be undefined and the SO client will
+          // use the current space.
+          initialNamespaces = [namespaceFallback];
+        }
+        return {
+          type: ML_TRAINED_MODEL_SAVED_OBJECT_TYPE,
+          id: m.model_id,
+          attributes: m,
+          ...(initialNamespaces ? { initialNamespaces } : {}),
+        };
+      })
+    );
+  }
+
+  async function getAllTrainedModelObjectsForAllSpaces(modelIds?: string[]) {
+    await isMlReady();
+    const searchFields = ['model_id'];
+    let filter = '';
+
+    if (modelIds !== undefined && modelIds.length) {
+      filter = modelIds
+        .map((m) => `${ML_TRAINED_MODEL_SAVED_OBJECT_TYPE}.attributes.model_id: "${m}"`)
+        .join(' OR ');
+    }
+
+    const options: SavedObjectsFindOptions = {
+      type: ML_TRAINED_MODEL_SAVED_OBJECT_TYPE,
+      perPage: 10000,
+      ...(spacesEnabled === false ? {} : { namespaces: ['*'] }),
+      searchFields,
+      filter,
+    };
+
+    return (await internalSavedObjectsClient.find<TrainedModelObject>(options)).saved_objects;
+  }
+
+  async function _deleteTrainedModel(modelId: string) {
+    const [model] = await _getTrainedModelObjects(modelId);
+    if (model === undefined) {
+      throw new MLJobNotFound('job not found');
```

Comment:
Do you need a new `MLModelNotFound` error type here, and the text needs to say `model not found` doesn't it?